### ISSUE TITLE
ACTION_A should be dispatched before ACTION_B

### ISSUE THREAD

**Issue by jordanmkoncz** [diegohaz/redux-saga-thunk#6]
First of all, great library, it's been really useful in allowing my components that dispatch an action to perform some additional functionality based on the eventual response from `redux-saga`.

I'm getting this error thrown from the `redux-saga-thunk` middleware and I think my usage of `redux-saga-thunk` is normal/valid and shouldn't result in any errors. The error is being thrown at https://github.com/diegohaz/redux-saga-thunk/blob/master/src/middleware.js#L26.

I have a standard action being dispatched, `ACTION_A`, with `meta: { thunk: true }`. I have a standard `redux` store configuration (matches the `redux-saga-thunk` "Basic setup" instructions) and a standard `redux-saga` setup, with my sagas essentially looking like this:

```
// sagas.js

function* handleActionA(action) {
  yield put({ type: 'ACTION_B', meta: action.meta });
}

function* watchActionA() {
  yield takeEvery('ACTION_A', handleActionA);
}

function* rootSaga() {
  yield fork(watchActionA);
}

export default rootSaga;
```

This is obviously simplified, but this is essentially what I'm doing. In my case, if `handleActionA` determines that some value has already been fetched from the API and is available in the `redux` store, then it just dispatches `ACTION_B`. So the result of this is that `ACTION_A` is dispatched from a component in my app, and this is picked up by the `watchActionA` saga, and then in `handleActionA` I immediately dispatch `ACTION_B`.

`redux-saga-thunk` then throws the error "[redux-saga-thunk] ACTION_A should be dispatched before ACTION_B". But this doesn't make sense, because `ACTION_A` *was* dispatched before `ACTION_B`, and with my saga setup, the only way that `ACTION_B` can be dispatched is after a saga `take`'s an `ACTION_A` and runs another function which then dispatches `ACTION_B`, so it's actually *impossible* for `ACTION_B` to be dispatched before `ACTION_A` (which the current wording of the error message suggests is happening).

This seems to be some sort of race condition that occurs when the success/failure action is dispatched too quickly after the initial request action? I was able to avoid this error being thrown by adding `yield delay(100);` right before `yield put({ type: 'ACTION_B', meta: action.meta });` in `handleActionA` but that is obviously not a good solution.

**Comment by diegohaz** [comment]
I've seen that error sometimes in my projects, but every time it was `ACTION_B` being dispatched twice after `ACTION_A`. So, the first time it has consumed `ACTION_A` and then there was no request action for the second `ACTION_B`, making it to throw.

That often happened to me when I got some error between, for example, `ACTION_SUCCESS` and the end of the saga:
```js
function* worker(action) {
  try {
    yield call(...)
    yield put('ACTION_SUCCESS', action.meta)
    throw new Error() // some error occurs here
  } catch (e) {
    yield put('ACTION_FAILURE', action.meta)
  }
}
```
That will dispatch `ACTION_SUCCESS`, but also `ACTION_FAILURE`, and then it will throw because `ACTION_REQUEST` has been already consumed by the successful action.

That's just an example that often occurs to me. If that's not your case, you should look at redux devtools to see if it has some weird stuff.

**Comment by jordanmkoncz** [comment]
Hey @diegohaz. That's a good suggestion and definitely something to look out for. In my case I don't believe that's what's happening.

I tested to make sure this wasn't happening by stripping down my code to essentially match the simplified `sagas.js` code I provided before. So my `handleActionA` function is exactly like the one shown in my example, it just dispatches `ACTION_B` immediately and there is no other code in the function. With the code like this, `ACTION_A` is dispatched, this is picked up by `watchActionA` and then `handleActionA` dispatches `ACTION_B`, and the "ACTION_A should be dispatched before ACTION_B" error is thrown. In the Redux devtools I can see `ACTION_A`, and it all looks correct, but `ACTION_B` is not visible in the devtools, I guess because the error is thrown before `ACTION_B` actually gets dispatched.

And I also again tested adding `yield delay(100);` to the beginning of `handleActionA`, and like before, this 100ms delay actually makes it so that `redux-saga-thunk` does not throw any error and everything works the way I expect it to.

So I'm fairly certain it's not a case of one action being dispatched which consumes the thunk/promise created by the initial action and then another action being dispatched which also tries to consume the same thunk/promise. With the code I'm using I don't believe it's possible for that to be happening. And given that adding a simple delay fixes the issue, it seems like this error is being thrown due to some sort of race condition.

**Comment by diegohaz** [comment]
Oh, I can reproduce it by removing this line from tests: https://github.com/diegohaz/redux-saga-thunk/blob/master/test/index.test.js#L12

Very good catch, @jordanmkoncz

**Comment by diegohaz** [comment]
Solution published: `v0.5.1`

**Comment by jordanmkoncz** [comment]
Nice work @diegohaz, looks like a good solution.
